src/bin/cargo/commands/clean.rs

```diff
@@ -17,29 +21,243 @@ pub fn cli() -> Command {
         .arg(
             flag(
                 "dry-run",
-                "Display what would be deleted without deleting anything",
+                "Display what would be deleted without deleting anything (unstable)",
             )
             .short('n'),
         )
+
+        // NOTE: Not all of these options may get stabilized. Some of them are
+        // very low-level details, and may not be something typical users need.
+        .arg(
+            optional_opt(
+                "gc",
+                "Delete old and unused files (unstable) (comma separated): all, download, target, shared-target",
+            )
+            .hide(true)
+            .value_name("KINDS")
+            .value_parser(
+                PossibleValuesParser::new(["all", "download", "target", "shared-target"]).map(|x|
+                    match x.as_str() {
+                        "all" => AutoGcKind::All,
+                        "download" => AutoGcKind::Download,
+                        "target" => panic!("target is not yet implemented"),
+                        "shared-target" => panic!("shared-target is not yet implemented"),
```

Comment:
Should we just not mention the "not implemented" stuff?

There are other things not implemented as well (e.g. cargo script cache) and we don't really know what the CLI will look like yet

Reply:
I removed the shared-target stuff, since that was a little far-fetched. However, I would like to keep the "target" part to indicate how it is intended to be used, and I think helps indicate how the pieces are intended to fit together. Otherwise, various parts I think would not really be clear as to why it is done like it is. I also don't expect the "target" part to be as far out, and has a rough plan for getting there.

Reply:
How do you expect to show these not-implemented parts?
- No behavior implemented, so interaction doesn't work
- They are hidden in help, so people can't browse it
- People can construct command-lines and see them panic but they can just construct them and see them fail with clap messages.

What if `clean::cli` checked the channel and only hid them if its stable?

Reply:
#6102 is somewhat related to this.

Reply:
I had either two intentions:
* Anything unimplemented would remain hidden when anything else is stabilized. The only people who would encounter it is someone reading the code.
* Unimplemented parts would get removed when moving towards stabilization if they are in the way or causing a hassle.

A big part of the path for this feature is figuring out what the CLI would look like. I am expecting the CLI to change, so I didn't want to put too much effort into polishing it right now.

Reply:
Not quite what I was getting at.  The unimplemented flags are here to "to indicate how it is intended to be used".  I originally took this to mean at the CLI level but at the CLI level they aren't exposed.  Do you instead mean at `ops` level?  If so, I don't feel its all that needed.  While the CLI isn't finalized and so we can let everything go, I feel like adding a bunch of code muddies the waters and made this review much bigger than it should be (even if doing the work helped to work around design issues).

Reply:
The "how it is intended to be used" was in regards to how the interface of `ops::clean()` works. So, for example, when cargo is instructed to clean caches, the interface of `ops::clean()` needs a variety of options to tell it what to clean. Those options include global cache, target cache, and potentially other caches as well. That's the purpose of the `AutoGcKind` enum. It is intended to show why `AutoGcKind` exists, how the detailed options interact, and how it all fit together, why it exists, how it is intended to evolve, etc.

I'm not sure I see it being a bunch of code. I see maybe 30 lines total in this file, which seems like a relatively minor part? Or maybe I'm not understanding or seeing what the concern is.

My concern is that removing the parts that this is intended to have will make it not as clear why it is structured the way it is, and the removed parts will just need to be re-added again. Keeping them here makes it easier to maintain in my opinion.

Reply:
If this is meant to show the interface for `ops::clean`, I think this then comes back to our other conversation about fundamentals vs details in UX (https://github.com/rust-lang/cargo/pull/12634#discussion_r1378171958).  For me, since I view the UX under review to be a `yeet`-like CLI that exists for the sake of exercising the core, it seem weird for that CLI to go "deeper".